### ISSUE TITLE
Windows Installer and Osquery Default Paths

### ISSUE THREAD

**Issue by jemos** [osquery/osquery#6098]
# Bug report

### What operating system and version are you using?

Windows

### What version of osquery are you using?

v4.1.1 from osquery.io site

### What steps did you take to reproduce the issue?

Just make a fresh install.

### What did you expect to see?

Osquery using default paths pointing to the installed directory. One of:

1. osquery using as default root path c:\osquery and installer installing to c:\osquery
2. osquery installer installing osquery to c:\program files\osquery and osqueryd using c:\program files\osquery as default path (as it is now)

### What did you see instead?

Osquery installer path and osqueryd default path seem to be inconsistent.

**Comment by theopolis** [comment]
@muffins, any idea what's going on?

**Comment by muffins** [comment]
@jemos interesting, not sure how that might've happened. If I pull down the osquery MSI from the site on a Windows install with no osquery deployed, I see the expected behavior:

Checking there's no previous osquery installed in any of the aforementioned locations:
```
PS C:\WINDOWS\system32> ls 'C:\Program Files\' | sls osquery                                                                                                        PS C:\WINDOWS\system32> sc.exe qc osqueryd                                                                                                                          [SC] OpenService FAILED 1060:

The specified service does not exist as an installed service.

PS C:\WINDOWS\system32> ls 'C:\' | sls osquery                                                                                                                      PS C:\WINDOWS\system32> 
```

If I then grab the MSI from `osquery.io` and install, here's what I see:
```
PS C:\WINDOWS\system32> ls 'C:\' | sls osquery                                                                                                                      PS C:\WINDOWS\system32> ls 'C:\Program Files\' | sls osquery                                                                                                        
osquery


PS C:\WINDOWS\system32> sc.exe qc osqueryd                                                                                                                          [SC] QueryServiceConfig SUCCESS

SERVICE_NAME: osqueryd
        TYPE               : 10  WIN32_OWN_PROCESS
        START_TYPE         : 2   AUTO_START
        ERROR_CONTROL      : 1   NORMAL
        BINARY_PATH_NAME   : "C:\Program Files\osquery\osqueryd\osqueryd.exe" --flagfile="C:\Program Files\osquery\osquery.flags"
        LOAD_ORDER_GROUP   :
        TAG                : 0
        DISPLAY_NAME       : osqueryd
        DEPENDENCIES       :
        SERVICE_START_NAME : NT AUTHORITY\SYSTEM
```

Can you offer how you're invoking the MSI? Have you tried uninstalling osquery via the Windows Program manager and then reinstalling? Any other insight you might be able to offer on your environment to help me re-create?

**Comment by theopolis** [comment]
Do we expect C:\osquery to be created by the installer? As far as I understand this path is not used and should not exist.

**Comment by directionless** [comment]
I'm not totally sure I understand what's happening.

@muffins Are you confirming that osquery is installing to `C:\osquery`, but that the service is using `C:\Program Files\osquery`? My first thought is that we should be in Program Files, and not root.

**Comment by zwass** [comment]
I just tested and after installing the MSI I **do not** have a `C:\osquery` directory.

```
PS E:\coveware-osquery> ls C:\osquery
ls : Cannot find path 'C:\osquery' because it does not exist.
At line:1 char:1
+ ls C:\osquery
+ ~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\osquery:String) [Get-ChildItem], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetChildItemCommand
```

I can confirm that the service does use the correct path.

```
PS E:\coveware-osquery> sc.exe qc osqueryd
[SC] QueryServiceConfig SUCCESS

SERVICE_NAME: osqueryd
        TYPE               : 10  WIN32_OWN_PROCESS
        START_TYPE         : 2   AUTO_START
        ERROR_CONTROL      : 1   NORMAL
        BINARY_PATH_NAME   : "C:\Program Files\osquery\osqueryd\osqueryd.exe" --flagfile="C:\Program Files\osquery\osquery.flags"
        LOAD_ORDER_GROUP   :
        TAG                : 0
        DISPLAY_NAME       : osqueryd
        DEPENDENCIES       :
        SERVICE_START_NAME : NT AUTHORITY\SYSTEM
PS E:\coveware-osquery> ls "C:\Program Files\osquery\osqueryd\osqueryd.exe"


    Directory: C:\Program Files\osquery\osqueryd


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       11/23/2019  11:49 PM       10992632 osqueryd.exe
```

@jemos can you please clarify what you mean by "Osquery installer path and osqueryd default path seem to be inconsistent."?

@muffins Your pasted powershell looks inconsistent... I don't see how you could have got `The specified service does not exist as an installed service.` as the result of the `ls` command you ran.

**Comment by jemos** [comment]
Sorry for the delay.
It seems I can only see this erratic behavior if I try to install it through the command line using msiexec. I'm trying to make a list of steps to reproduce in my environment.

**Comment by jemos** [comment]
Meanwhile I've a log file of the MSI installer, and there we can see that:

```
Action start 8:09:11: FileCost.
MSI (s) (30:9C) [08:09:11:953]: Doing action: CostFinalize
MSI (s) (30:9C) [08:09:11:953]: Note: 1: 2205 2:  3: ActionText 
Action ended 8:09:11: FileCost. Return value 1.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding OutOfDiskSpace property. Its value is '0'.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding OutOfNoRbDiskSpace property. Its value is '0'.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding PrimaryVolumeSpaceAvailable property. Its value is '0'.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding PrimaryVolumeSpaceRequired property. Its value is '0'.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding PrimaryVolumeSpaceRemaining property. Its value is '0'.
MSI (s) (30:9C) [08:09:11:953]: Note: 1: 2205 2:  3: Patch 
MSI (s) (30:9C) [08:09:11:953]: Note: 1: 2205 2:  3: Condition 
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding TARGETDIR property. Its value is 'C:\'.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Modifying ProgramFiles64Folder property. Its current value is 'C:\Program Files\'. Its new value: 'C:\'.    <-------
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding INSTALLFOLDER property. Its value is 'C:\osquery\'.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding FileSystemLogging property. Its value is 'C:\osquery\log\'.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding CertsFolder property. Its value is 'C:\osquery\certs\'.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding PacksFolder property. Its value is 'C:\osquery\packs\'.
MSI (s) (30:9C) [08:09:11:953]: PROPERTY CHANGE: Adding DaemonFolder property. Its value is 'C:\osquery\osqueryd\'.
MSI (s) (30:9C) [08:09:11:953]: Target path resolution complete. Dumping Directory table...
MSI (s) (30:9C) [08:09:11:953]: Note: target paths subject to change (via custom actions or browsing)
MSI (s) (30:9C) [08:09:11:953]: Dir (target): Key: TARGETDIR	, Object: C:\
MSI (s) (30:9C) [08:09:11:953]: Dir (target): Key: ProgramFiles64Folder	, Object: C:\
MSI (s) (30:9C) [08:09:11:953]: Dir (target): Key: INSTALLFOLDER	, Object: C:\osquery\
MSI (s) (30:9C) [08:09:11:953]: Dir (target): Key: FileSystemLogging	, Object: C:\osquery\log\
MSI (s) (30:9C) [08:09:11:953]: Dir (target): Key: CertsFolder	, Object: C:\osquery\certs\
MSI (s) (30:9C) [08:09:11:953]: Dir (target): Key: PacksFolder	, Object: C:\osquery\packs\
MSI (s) (30:9C) [08:09:11:953]: Dir (target): Key: DaemonFolder	, Object: C:\osquery\osqueryd\
```

**Comment by jemos** [comment]
To reproduce the issue as in my environment, please do:
1. Start `cmd.exe` shell
2. Using `psexec.exe` from SysInternals, get a `SYSTEM` user cmd with: `psexec -i -s cmd.exe`
3. In the new cmd call powershell to install with:

```powershell.exe -Command "& {Set-ExecutionPolicy Bypass -Scope Process -Force; msiexec.exe /quiet /a osquery-4.1.1.msi /L*V c:\osquery_install.log}"```

In the end, dir c:\ will have:

```
c:\Windows\Temp>dir c:\
dir c:\
 Volume in drive C has no label.
 Volume Serial Number is C840-7F30

 Directory of c:\

11/29/2019  04:16 PM    <DIR>          Go
12/30/2019  09:59 AM    <DIR>          osquery    <-- created by installer
12/30/2019  09:59 AM           262,144 osquery-4.1.1.msi    <-- created by installer
12/30/2019  09:59 AM            90,982 osquery_install.log
```

**Comment by Smjert** [comment]
@jemos It seems that using `/a` does an Administrative Install which is basically an extract operation, more than a normal install, and which can also accept a TARGETDIR parameter.
Is there a specific reason why you're not installing with `/i`?

**Comment by jemos** [comment]
Well.. I thought it would install the package using *administrative privileges*, my bad. So, no particular reason to use it. I did some trial and error to get this `msiexec` to install correctly and I ended up using that `/a`.

I checked and indeed, using:
- `/a` messes the installation and creates both `c:\osquery` and `c:\program files\osquery` one with binaries other with configuration files.
- `/i` installs only in `c:\program files\osquery`

Thanks and sorry for your time spent on this... it seems it was just a problem of the way I was installing.
